File: LeetCode/46.permutations/46.Permutations.py
#!/usr/bin/env python3


class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        res = []
        n = len(nums)

        def backtrack(i):
            # print(nums)
            if i == n:
                res.append(nums[:])
                return

            for j in range(i, n):
                nums[i], nums[j] = nums[j], nums[i]
                backtrack(i + 1)
                nums[i], nums[j] = nums[j], nums[i]

        backtrack(0)

        return res

File: LeetCode/46.permutations/46.Permutations.swift
class Solution {
    func permute(_ nums: [Int]) -> [[Int]] {
        let n = nums.count
        var nums = nums
        var res: [[Int]] = []

        func backtrack(_ i: Int) {
            if i == nums.count {
                res.append(nums)
                return
            }

            for j in i ..< n {
                nums.swapAt(j, i)
                backtrack(i + 1)
                nums.swapAt(i, j)
            }
        }

        backtrack(0)

        return res
    }
}
